Bookmarks model not loaded

AlexeyBarabash opened the pull request:
<!-- Add brave-browser issue bellow that this PR will resolve -->
Resolves https://github.com/brave/brave-browser/issues/8178

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues/8178) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [x] Linux
  - [ ] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [x] Linux
  - [ ] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [ ] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:
See in comments below 

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

AlexeyBarabash commented:
### Test plan

Here is a way how to emulate such case.
1. Revert PR https://github.com/brave/brave-core/pull/4565
2. Build browser
3. Launch profiles A and B and enable chain between them
4. Close browsers of profiles A and B
5. Put huge delay into `model_loader.cc` and build browser like below
```
git diff browser/model_loader.cc
diff --git a/components/bookmarks/browser/model_loader.cc b/components/bookmarks/browser/model_loader.cc
index 12fe8d32547f..b77e87ec8180 100644
--- a/components/bookmarks/browser/model_loader.cc
+++ b/components/bookmarks/browser/model_loader.cc
@@ -10,7 +10,8 @@
 #include "base/threading/thread_task_runner_handle.h"
 #include "components/bookmarks/browser/bookmark_storage.h"
 #include "components/bookmarks/browser/url_index.h"
-
+  #include "base/synchronization/waitable_event.h"
+  #include "base/threading/thread_restrictions.h"
 namespace bookmarks {
 
 namespace {
@@ -62,9 +63,17 @@ void ModelLoader::DoLoadOnBackgroundThread(
     scoped_refptr<base::SequencedTaskRunner> main_sequenced_task_runner,
     std::unique_ptr<BookmarkLoadDetails> details,
     LoadCallback callback) {
   LoadBookmarks(profile_path, emit_experimental_uma, details.get());
   history_bookmark_model_ = details->url_index();
+  const base::TimeDelta kExecutionTime(base::TimeDelta::FromMilliseconds(20*1000));
+  base::ScopedAllowBaseSyncPrimitivesForTesting allow_base_sync_primitives;
+  base::WaitableEvent hang_event_;
+  hang_event_.TimedWait(kExecutionTime);
   loaded_signal_.Signal();
   main_sequenced_task_runner->PostTask(
       FROM_HERE, base::BindOnce(&ModelLoader::OnFinishedLoad, this,
                                 std::move(details), std::move(callback)));
@@ -72,6 +81,7 @@ void ModelLoader::DoLoadOnBackgroundThread(
 
 void ModelLoader::OnFinishedLoad(std::unique_ptr<BookmarkLoadDetails> details,
                                  LoadCallback callback) {
   std::move(callback).Run(std::move(details));
 }
```
6. Launch profile A
7. See crash on pre-PR master

bridiver commented on components/brave_sync/brave_profile_sync_service_impl.h:
```diff
@@ -215,6 +215,16 @@ class BraveProfileSyncServiceImpl
+  // Methods to run specified actions right after bookmark model was loaded
+  void RunWhenModelIsLoaded(base::OnceCallback<void()> fn,
+                            const std::string& fn_name);
+  void CtorBkmLoaded(Profile* profile);
```
please don't abbreviate things like this in method names. We should use the same terminology everywhere for consistency

AlexeyBarabash replied:
thanks, now this is fixed

bridiver commented on components/brave_sync/brave_profile_sync_service_impl.cc:
```diff
@@ -280,6 +357,14 @@ void BraveProfileSyncServiceImpl::OnSetupSyncNewToSync(
 
   brave_sync_initializing_ = true;
 
+  RunWhenModelIsLoaded(
```
I'm concerned about running parts of these methods, but not completing them. I think it's likely to introduce subtle bugs. I think it would be better to just wait on everything so we'd have `OnSetupSyncNewToSync` that would call `OnSetupSyncNewToSyncImpl` when the bookmark model is loaded
```

AlexeyBarabash replied:
thanks, now this is redone at https://github.com/brave/brave-core/pull/4588/commits/361b930cb4df1405b5f89d10e5673bd061f7b7db

bridiver commented on components/brave_sync/brave_profile_sync_service_impl.cc:
```diff
@@ -1042,6 +1145,18 @@ BraveSyncService* BraveProfileSyncServiceImpl::GetSyncService() const {
     const std::string& category_name,
     RecordsListPtr records) {
+  RunWhenModelIsLoaded(
+      base::BindOnce(&BraveProfileSyncServiceImpl::SendSyncRecordsBkmLoaded,
```
the convention here is `SendSyncRecordsImpl` and again we don't want to use abbreviations like Bkm

darkdh commented on components/brave_sync/brave_profile_sync_service_impl.h:
```diff
@@ -215,6 +221,13 @@ class BraveProfileSyncServiceImpl
 
+  // Methods to run specified actions right after bookmark model was loaded
+  void OnSetupSyncNewToSyncBkmLoaded();
+  void OnSetupSyncHaveCodeBkmLoaded(const Uint8Array& seed);
```
we shouldn't queue `OnSetupSyncNewToSync` and `OnSetupSyncHaveCode` because both have to wait `OnSyncReady` so we only need to queue up `OnSyncReady``

darkdh commented on components/brave_sync/brave_profile_sync_service_impl.h:
```diff
@@ -215,6 +221,13 @@ class BraveProfileSyncServiceImpl
+  void OnSetupSyncNewToSyncBkmLoaded();
+  void OnSetupSyncHaveCodeBkmLoaded(const Uint8Array& seed);
+  void OnSyncReadyBkmLoaded();
+  void SendSyncRecordsBkmLoaded(const std::string& category_name,
```
we shouldn't queue `SendSyncRecords` either, if bookmark model is not loaded, commit message won't be constructed and `OnNudgeSyncCycle` won't be triggered

darkdh commented on components/brave_sync/brave_profile_sync_service_impl.h:
```diff
@@ -10,6 +10,8 @@
 #include <string>
 #include <vector>
 
+#include "base/one_shot_event.h"
```
seems like it is not needed anymore

AlexeyBarabash replied:
thanks, fixed at https://github.com/brave/brave-core/pull/4588/commits/faae417c8590f9732feec99c5b4df889845ab060

darkdh commented on components/brave_sync/brave_profile_sync_service_impl.h:
```diff
@@ -255,10 +264,14 @@ class BraveProfileSyncServiceImpl
   // this object was created on.
   SEQUENCE_CHECKER(sequence_checker_);
 
+  base::WeakPtrFactory<BraveProfileSyncServiceImpl> weak_ptr_factory_{this};
```
seems like it is not needed anymore

AlexeyBarabash replied:
thanks, fixed at https://github.com/brave/brave-core/pull/4588/commits/faae417c8590f9732feec99c5b4df889845ab060

darkdh commented on components/brave_sync/brave_profile_sync_service_impl.cc:
```diff
@@ -10,6 +10,7 @@
 #include <vector>
 
 #include "base/bind.h"
+#include "base/callback.h"
```
seems like it is not needed anymore

AlexeyBarabash replied:
thanks, fixed at https://github.com/brave/brave-core/pull/4588/commits/faae417c8590f9732feec99c5b4df889845ab060

darkdh commented on components/brave_sync/brave_sync_service_unittest.cc:
```diff
@@ -826,6 +826,20 @@ TEST_F(BraveSyncServiceTest, OnSyncReadyNewToSync) {
 }
 
+TEST_F(BraveSyncServiceTest, OnSyncReadyModelNotYetLoaded) {
+  EXPECT_NE(sync_service()->model_, nullptr);
```
I think we have better way to test if model is truly loaded not just nullptr
see `components/bookmarks/test/test_bookmark_client.cc`, we can have our own create method but not calling `DoneLoading` and then after calling `OnSyncReady`, we call `DoneLoading`

AlexeyBarabash replied:
cool, thanks for the advise, will redo

AlexeyBarabash replied:
redone, though I had to add patch  to access private `BookmarkModel::DoneLoading`.
cc @bridiver

AlexeyBarabash commented:
Updated tests and rebased.

darkdh commented on components/brave_sync/brave_sync_service_unittest.cc:
```diff
@@ -246,6 +250,38 @@ class BraveSyncServiceTest : public testing::Test {
+    model()->DoneLoading(std::move(bookmark_load_details_));
+  }
+ private:
+  static std::unique_ptr<KeyedService> BuildModel(content::BrowserContext* context) {
```
why is it a static function here? it access `bookmark_load_details_` and it is even in `private` section

AlexeyBarabash replied:
My, bad. I was sure `KeyedServiceFactory::SetTestingFactory` takes plain function as a parameter, but it takes `base::RepeatingCallback`. Fixed that, thanks for pointing.

darkdh commented on patches/components-bookmarks-browser-bookmark_model.h.patch:
```diff
@@ -0,0 +1,12 @@
+   friend class BookmarkStorage;
+   friend class ScopedGroupBookmarkActions;
+   friend class TestBookmarkClient;
++  BRAVE_FRIEND_CLASS;
```
The macro should be `BRAVE_BOOKMARK_MODEL_H` and insert before `private:`
see `BRAVE_BROWSER_H`

AlexeyBarabash replied:
thanks, fixed

AlexeyBarabash commented:
CI failed for iOS
```
loning into '/Users/jenkins/jenkins/workspace/ld-pr_bookmarks_model_not_loaded/src/_gclient_brave_qXZE2t'...
12:43:39  2>done.
12:43:55  Error: Command 'git ls-tree HEAD third_party/libgav1/src' returned non-zero exit status 128 in /Users/jenkins/jenkins/workspace/ld-pr_bookmarks_model_not_loaded/src
12:43:55  fatal: not a tree object
```

AlexeyBarabash commented:
Have failed tests on CI
```
15:52:22  2 tests failed:
15:52:22      AutoplayPermissionContextBrowserTest.AskAutoplay (../../brave/browser/autoplay/autoplay_permission_context_browsertest.cc:227)
15:52:22      BraveRewardsBrowserTest.MultipleTipsProduceMultipleFeesWithVerifiedWallet (../../brave/components/brave_rewards/browser/rewards_service_browsertest.cc:3021)
```

AlexeyBarabash commented:
Failed browser test on Linux 
```

19:50:53  1 test failed:
19:50:53      AutoplayPermissionContextBrowserTest.ClickBlock (../../brave/browser/autoplay/autoplay_permission_context_browsertest.cc:340)
```
this is known problem

bsclifton commented:
@darkdh @AlexeyBarabash does this need to be in the hotfix we're planning for Wednesday?

cc: @kjozwiak @rebron

AlexeyBarabash commented:
@bsclifton yes, I would like to have this as a hotfix because it fixes one of sync crashes. From the other side, sync is disabled by default now, so maybe it is not uber-urgent. And this PR blocks one of @darkdh sync PRs (https://github.com/brave/brave-core/pull/4404)

darkdh commented:
This is also a delay startup crash for some users when starting up to get sync ready

bridiver commented on components/brave_sync/bookmark_model_loaded_only_observer.h:
```diff
@@ -0,0 +1,65 @@
+
+// This is a wrapper for the BookmarkModelObserver if we need only
+// BookmarkModelLoaded notification
+class BookmarkModelLoadedOnlyObserver
```
I'm not understanding the point of this class. This is even more verbose than just implementing empty methods on BraveProfileSyncServiceImpl

AlexeyBarabash replied:
@bridiver  because just 1 of 8 methods is used and 7 of 8 are not used in `BraveProfileSyncServiceImpl`. I had the idea to avoid increasing of `brave_profile_sync_service_impl.h` and to make it easier read when there is one method which is used then 7 which are not

bridiver replied:
I just don't think it's worth it to add another class just for that and also it makes things harder to understand imo. I had to reference `BookmarkModelLoadedOnlyObserver` because it wasn't obvious why it didn't subclass `BookmarkModelObserver`.

bridiver replied:
Introducing new classes just to stub methods only increases the load on people who are trying to read the code because it's one more new thing they have to learn. If it was more generally useful then there might be a better argument for it, but that isn't the case here since it's only used in one class.

AlexeyBarabash replied:
this was fixed at https://github.com/brave/brave-core/pull/4588/files#diff-4718e5fe0385d7ab2103a3e905d5fb09R77

bridiver commented on components/brave_sync/brave_profile_sync_service_impl.cc:
```diff
@@ -219,6 +230,9 @@ void BraveProfileSyncServiceImpl::OnNudgeSyncCycle(RecordsListPtr records) {
 
 BraveProfileSyncServiceImpl::~BraveProfileSyncServiceImpl() {
   network_connection_tracker_->RemoveNetworkConnectionObserver(this);
+  if (is_model_loaded_observer_set_) {
```
you don't have to check before removing. Removing an observer that doesn't exist is a no-op

bridiver replied:
if this flag is still needed for testing then it should be renamed to `is_model_loaded_observer_set_for_testing`

bridiver replied:
but it seems like it would be better to test with a mock observer instead

AlexeyBarabash replied:
thanks, I recall, I got failed DCHECK crash one time.
I would be happy to get rid of `is_model_loaded_observer_set_`

AlexeyBarabash replied:
this was fixed at https://github.com/brave/brave-core/pull/4588/files#diff-ab344835b09fb6ac1c601af787b116bbR230

bridiver commented on components/brave_sync/brave_profile_sync_service_impl.cc:
```diff
@@ -470,6 +484,19 @@ void BraveProfileSyncServiceImpl::OnSyncReady() {
+    // Will call OnSyncReadyBookmarksModelLoaded once model is loaded
+    LOG(WARNING) << "[BraveSync] bookmarks model is not yet loaded, "
+                 << "OnSyncReady will be delayed";
+    model_->AddObserver(this);
```
why aren't we just adding the observer in the constructor?

AlexeyBarabash replied:
because usually model is loaded before the sync service starts. in fact I had to put the `wait` into STR to be able to reproduce it. So observer is set when it is required and observer is not set when we are good without it.

bridiver commented on components/brave_sync/brave_sync_service_unittest.cc:
```diff
@@ -246,6 +250,37 @@ class BraveSyncServiceTest : public testing::Test {
 
+class BraveSyncServiceTestDelayedLoadModel : public BraveSyncServiceTest {
+ public:
+  void PrepareModelFactory() override {
```
nit: I think `SetBookmarkModelFactory` would make more sense here

AlexeyBarabash replied:
I was hesitant between `Set` and `Prepare`, so will change to Set.

AlexeyBarabash replied:
this was fixed at https://github.com/brave/brave-core/pull/4588/files#diff-8b38d039014ebfe5a6c917bf6ee86992R216

bridiver commented on components/brave_sync/brave_profile_sync_service_impl.cc:
```diff
@@ -200,11 +200,22 @@ BraveProfileSyncServiceImpl::BraveProfileSyncServiceImpl(Profile* profile,
 
+void BraveProfileSyncServiceImpl::BookmarkModelLoaded(BookmarkModel* model,
+                                                      bool ids_reassigned) {
+  DCHECK(model);
```
are these DCHECKs really necessary? `BookmarkModelLoaded` only gets called when the bookmark model is loaded so this would actually be debugging internal chromium code which doesn't seem necessary here

AlexeyBarabash replied:
>are these DCHECKs really necessary?
no that is my paranoia, if can't be null, but if it would be null, than all is totally wrong

AlexeyBarabash replied:
Fixed at https://github.com/brave/brave-core/pull/4588/files#diff-ab344835b09fb6ac1c601af787b116bbR211

bridiver commented on components/brave_sync/brave_profile_sync_service_impl.cc:
```diff
@@ -200,11 +200,22 @@ BraveProfileSyncServiceImpl::BraveProfileSyncServiceImpl(Profile* profile,
+  DCHECK(model);
+  DCHECK(model->loaded());
+
+  LOG(WARNING) << "[BraveSync] bookmarks model just loaded, "
```
this should be a vlog, not a warning

AlexeyBarabash replied:
fixed at https://github.com/brave/brave-core/pull/4588/files#diff-ab344835b09fb6ac1c601af787b116bbR211 and https://github.com/brave/brave-core/pull/4588/files#diff-ab344835b09fb6ac1c601af787b116bbR487

AlexeyBarabash commented:
@bridiver , I have addressed your comments, could please re-review?

AlexeyBarabash commented:
CI failed because of version mismatch though version in b-c was correct; deleted branch `bookmarks_model_not_loaded` at b-b and restarted ci

darkdh commented:
```
0   libbase.dylib                       0x0000000107c76259 base::debug::CollectStackTrace(void**, unsigned long) + 9
1   libbase.dylib                       0x0000000107b31b63 base::debug::StackTrace::StackTrace() + 19
2   libbase.dylib                       0x0000000107c76101 base::debug::(anonymous namespace)::StackDumpSignalHandler(int, __siginfo*, void*) + 2385
3   libsystem_platform.dylib            0x00007fff6a4a542d _sigtramp + 29
4   ???                                 0x000000010bdbc668 0x0 + 4493919848
5   libbase.dylib                       0x0000000107baf7be base::SequenceCheckerImpl::CalledOnValidSequence() const + 30
6   brave_unit_tests                    0x0000000101fe9032 bookmarks::BookmarkModel::RemoveObserver(bookmarks::BookmarkModelObserver*) + 34
7   brave_unit_tests                    0x00000001010d83b3 brave_sync::BraveProfileSyncServiceImpl::~BraveProfileSyncServiceImpl() + 323
8   brave_unit_tests                    0x00000001010d859e brave_sync::BraveProfileSyncServiceImpl::~BraveProfileSyncServiceImpl() + 14
9   libkeyed_service_core.dylib         0x000000010b1e86e0 KeyedServiceFactory::Disassociate(void*) + 192
10  libkeyed_service_core.dylib         0x000000010b1e8762 KeyedServiceFactory::ContextDestroyed(void*) + 18
11  libkeyed_service_core.dylib         0x000000010b1e782c DependencyManager::PerformInterlockedTwoPhaseShutdown(DependencyManager*, void*, DependencyManager*, void*) + 668
12  brave_unit_tests                    0x00000001012c2a17 TestingProfile::~TestingProfile() + 119
13  brave_unit_tests                    0x00000001012c2e1e TestingProfile::~TestingProfile() + 14
```

```
5 tests crashed:
    AutoplayPermissionContextTests.TestNonAutoRefresh (../../brave/browser/autoplay/autoplay_permission_context_unittest.cc:126)
    BraveAutocompleteProviderClientUnitTest.SyncURLSuggestedNotSyncInternal (../../brave/browser/autocomplete/brave_autocomplete_provider_client_unittest.cc:52)
    BraveExtensionsAPIClientTests.IsBraveProtectedUrl (../../brave/browser/extensions/api/brave_extensions_api_client_unittest.cc:32)
    PushMessagingNotificationManagerTest.IsTabVisible (../../chrome/browser/push_messaging/push_messaging_notification_manager_unittest.cc:23)
    PushMessagingNotificationManagerTest.IsTabVisibleViewSource (../../chrome/browser/push_messaging/push_messaging_notification_manager_unittest.cc:41)
```

AlexeyBarabash commented:
Pushed fix for crashed tests.
https://github.com/brave/brave-core/compare/36acb17cbb745a5a48aede95c31c206c591309e6..c73e82987cb02554c4535f855e3faad47ec55f76

@bridiver could you look on that please, as this fix reverts on of your review notices, now we check again, whether the observer was set before remove.

AlexeyBarabash commented:
failed iOS build with 
```
08:59:31  ERROR at //build/timestamp.gni:31:19: Script returned non-zero exit code.
08:59:31  build_timestamp = exec_script(compute_build_timestamp,
08:59:31                    ^----------
08:59:31  Current dir: /Users/jenkins/jenkins/workspace/ld-pr_bookmarks_model_not_loaded/src/out/ios_Release/
08:59:31  Command: python /Users/jenkins/jenkins/workspace/ld-pr_bookmarks_model_not_loaded/src/build/compute_build_timestamp.py official
08:59:31  Returned 1.
08:59:31  stderr:
```
failed linux browser test 
```
09:48:54  ../../brave/browser/autoplay/autoplay_permission_context_browsertest.cc:510: Failure
09:48:54  Expected: (result) != (kVideoPlaying), actual: "Video playing" vs "Video playing"
09:48:54  Stack trace:
09:48:54  #0 0x56402bb1e209 (/home/ubuntu/jenkins/workspace/ld-pr_bookmarks_model_not_loaded/src/out/Release/brave_browser_tests+0x501c208)
....
09:50:51  1 test failed:
09:50:51      AutoplayPermissionContextBrowserTest.BlockAutoplay (../../brave/browser/autoplay/autoplay_permission_context_browsertest.cc:468)
```
failed MacOS build
```
10:07:58  In file included from ../../brave/browser/tor/tor_profile_service_impl.cc:14:
10:07:58  ../../brave/browser/brave_browser_process_impl.h:14:10: fatal error: 'brave/components/brave_referrals/buildflags/buildflags.h' file not found
10:07:58  #include "brave/components/brave_referrals/buildflags/buildflags.h"
10:07:58           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
10:07:58  1 error generated.
```

AlexeyBarabash commented:
CI failed because could not apply patch to version, maybe correct chromium src dir wasn't fetched.
```
14:19:14  1 failed patches:
14:19:14  chrome/VERSION
14:19:14    - Patch applied because: No corresponding .patchinfo file was found.
14:19:14    - Error - Program git exited with error code 1.
14:19:14  error: patch failed: chrome/VERSION:1
14:19:14  error: chrome/VERSION: patch does not apply
```
removed b-b branch and rebased
